meetings/working-groups/collection-literals/CL-2023-04-05.md

```diff
@@ -0,0 +1,94 @@
+# Collection literals working group meeting for April 4, 2023
+
+## Agenda
+
+* Allow efficient construction of collections that are typically represented as arrays, particularly `ImmutableArray<T>`.
```

Comment:
I felt like our conversation was also applicable to construction of mutable and immutable dictionaries and sets, etc.

Reply:
We could consider extending this to allow attributing factory methods for constructing mutable and immutable dictionaries, sets, etc. Would that be for efficiency or simply to enable constructing types that are currently constructible such as `ImmutableDictionary<TKey, TValue>`?

Reply:
Both for efficiency and flexibility.

I feel that we should investigate whether a single pattern like the following could address all the construction scenarios we care about now or may care about in the future, while also providing the high performance we want here:
1. `CreateBuilder(capacity)`--maybe with a hint about whether the exact size is known for sure, or if you may receive a different number of elements than the given capacity.
2. `Add()` 0 or more times
3. `BuildAndClear()`

The builder should be a struct to reduce the amount of allocations, but the language doesn't need to require that.

`Add()` should do a minimal amount of work, be inlinable, and not do things like update a counter or whatever else.

`BuildAndClear()` could be designed with the assumption that the builder is in an invalid state afterwards, and doesn't need to account for being called again.

I avoided the `ToSpecificCollectionType()` naming convention here because I want the method names to be fixed on conventions, like the way we do Deconstruct, GetPinnableReference, so on and so forth.

Reply:
> I feel that we should investigate whether a single pattern like the following could address all the construction scenarios we care about now or may care about in the future, while also providing the high performance we want here:

Sounds good. Let's add this to the agenda for the next meeting.

Reply:
i'd want @stephentoub  to weigh in, as that's basically requiring such a surface area be provided (as opposed to the approach we talked about in the meeting, which would generally just be a constructor, or a single static method with an attribute) :)

My concern is this being a large tax for the rare case.